src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,559 @@
+  bool _progress;
+  bool _present;
+  T _data;
+};
+
+template <class T>
+class NormalizeSimpleResult {
+public:
+  bool _present;
+  RangeInt<T> _bounds;
+  KnownBits<T> _bits;
+};
+
+// Try to tighten the bound constraints from the known bit information
+// E.g: if lo = 0 but the lowest bit is always 1 then we can tighten
+// lo = 1
+template <class T>
+static AdjustResult<RangeInt<T>>
+adjust_bounds_from_bits(const RangeInt<T>& bounds, const KnownBits<T>& bits) {
+  static_assert(std::is_unsigned<T>::value, "");
+
+  auto adjust_lo = [](T lo, const KnownBits<T>& bits) {
+    constexpr size_t W = sizeof(T) * 8;
+    T zero_violation = lo & bits._zeros;
+    T one_violation = ~lo & bits._ones;
+    if (zero_violation == one_violation) {
+      return lo;
+    }
+
+    if (zero_violation < one_violation) {
+      // Align the last violation of ones unset all the lower bits
+      // so we don't care about violations of zeros
+      juint last_violation = W - 1 - count_leading_zeros(one_violation);
+      T alignment = T(1) << last_violation;
+      lo = (lo & -alignment) + alignment;
+      return lo | bits._ones;
+    }
+
+    // Suppose lo = 00110010, zeros = 01010010, ones = 10001000
+    // Since the 4-th bit must be 0, we need to align up the lower bound.
+    // This results in lo = 01000000, but then the 6-th bit does not match,
+    // align up again gives us 10000000.
+    // We can align up directly to 10000000 by finding the first place after
+    // the highest mismatch such that both the corresponding bits are unset.
+    // Since all bits lower than the alignment are unset we don't need to
+    // align for the violations of ones anymore.
+    juint last_violation = W - 1 - count_leading_zeros(zero_violation);
+    T find_mask = std::numeric_limits<T>::max() << last_violation;
+    T either = lo | bits._zeros;
+    T tmp = ~either & find_mask;
+    T alignment = tmp & (-tmp);
+    lo = (lo & -alignment) + alignment;
+    return lo | bits._ones;
+  };
+
+  T new_lo = adjust_lo(bounds._lo, bits);
+  if (new_lo < bounds._lo) {
+    return {true, false, {}};
+  }
+  T new_hi = ~adjust_lo(~bounds._hi, {bits._ones, bits._zeros});
+  if (new_hi > bounds._hi) {
+    return {true, false, {}};
+  }
+
+  bool progress = (new_lo != bounds._lo) || (new_hi != bounds._hi);
+  bool present = new_lo <= new_hi;
+  return {progress, present, {new_lo, new_hi}};
+}
+
+// Try to tighten the known bit constraints from the bound information
+// E.g: if lo = 0 and hi = 10, then all but the lowest 4 bits must be 0
+template <class T>
+static AdjustResult<KnownBits<T>>
+adjust_bits_from_bounds(const KnownBits<T>& bits, const RangeInt<T>& bounds) {
+  static_assert(std::is_unsigned<T>::value, "");
+  T mismatch = bounds._lo ^ bounds._hi;
+  T match_mask = mismatch == 0 ? std::numeric_limits<T>::max()
+                               : ~(std::numeric_limits<T>::max() >> count_leading_zeros(mismatch));
+  T new_zeros = bits._zeros | (match_mask &~ bounds._lo);
```

Comment:
```suggestion
  T new_zeros = bits._zeros | (match_mask & ~bounds._lo);
```
I think this was a typo?

Reply:
It looks more like an `and not` to me :) However, if you prefer the `~` to stick to `bounds._lo` I would make that change.

Reply:
I was confused by it, and wondered if a `&~` operator exists in cpp... :sweat_smile:
But this is only a weak preference on my side.